File: stepik-union/src/org/stepik/plugin/actions/DirectivesUtils.java
package org.stepik.plugin.actions;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.stepik.plugin.collective.SupportedLanguages;

import java.util.Arrays;

public class DirectivesUtils {
    private static final String START_DIRECTIVE = "Stepik code: start";
    private static final String END_DIRECTIVE = "Stepik code: end";
    private static final String START_HINT = "Please note, only the code below will be sent to Stepik.org";
    private static final String END_HINT = "Please note, only the code above will be sent to Stepik.org";

    private static final String MESSAGE = "Do you want to remove Stepik directives and external code?\n" +
            "You can undo this action using \"ctrl + Z\".";

    public static String[] getFileText(VirtualFile vf) {
        Document document = FileDocumentManager.getInstance().getDocument(vf);
        return document != null ? document.getText().split("\n") : new String[0];
    }

    public static String getTextUnderDirectives(String[] text, SupportedLanguages lang) {
        Pair<Integer, Integer> locations = findDirectives(text, lang);

        int start = locations.first;
        int end = locations.second;

        StringBuilder sb = new StringBuilder();

        for (int i = start + 1; i < end; i++) {
            sb.append(text[i]).append("\n");
        }

        return sb.toString();
    }

    /**
     * Find first "Stepik code: start" and last "Stepik code: end". Return Pair(start, end).
     * <p>
     * If "Stepik code: start" not found, start = -1
     * <p>
     * If "Stepik code: end" not found, end = text.length
     */
    public static Pair<Integer, Integer> findDirectives(String[] text, SupportedLanguages lang) {
        int start = -1;
        int end = text.length;

        for (int i = 0; i < text.length; i++) {
            if (isStart(text[i], lang)) {
                start = i;
                break;
            }
        }

        for (int i = start + 1; i < text.length; i++) {
            if (isEnd(text[i], lang)) {
                end = i;
                break;
            }
        }

        return Pair.create(start, end);
    }

    private static boolean isStart(String line, SupportedLanguages lang) {
        if (!lang.isCommentedLine(line))
            return false;

        line = line.trim().substring(lang.getComment().length()).trim();

        return START_DIRECTIVE.equals(line);
    }

    private static boolean isEnd(String line, SupportedLanguages lang) {
        if (!lang.isCommentedLine(line))
            return false;

        line = line.trim().substring(lang.getComment().length()).trim();

        return END_DIRECTIVE.equals(line);
    }

    public static void writeInToFile(String[] text, VirtualFile file, Project project) {
        final Document document = FileDocumentManager.getInstance().getDocument(file);
        if (document == null)
            return;
        final StringBuilder sb = new StringBuilder();
        for (String tmp : text)
            sb.append(tmp).append("\n");

        CommandProcessor
                .getInstance()
                .executeCommand(project,
                        () -> ApplicationManager
                                .getApplication()
                                .runWriteAction(
                                        () -> document.setText(sb.toString())),
                        "Stepik directives process",
                        "Stepik directives process");
    }

    public static String[] removeAmbientCode(String[] text,
                                             Pair<Integer, Integer> locations,
                                             Project project,
                                             @NotNull SupportedLanguages lang) {
        int start = locations.first;
        int end = locations.second;

        String[] before = Arrays.copyOfRange(text, 0, start > 0? start : 0);
        String[] after = Arrays.copyOfRange(text, end + 1, text.length);

        if (!Arrays.equals(before, lang.getBeforeCode()) ||
                !Arrays.equals(after, lang.getAfterCode())) {
            int information = Messages.showYesNoDialog(project, MESSAGE, "Information", Messages.getInformationIcon());
            if (information != 0) return text;
        }

        String[] result = new String[end - (start + 1)];

        for (int i = 0; i < result.length; i++) {
            result[i] = text[i + start + 1];
        }

        return result;
    }

    public static String[] insertAmbientCode(@NotNull String[] text, @NotNull SupportedLanguages lang, boolean showHint) {
        String[] beforeCode = lang.getBeforeCode();
        String[] afterCode = lang.getAfterCode();

        int k = showHint ? 2 : 1;

        String[] ans = new String[text.length + beforeCode.length + afterCode.length + 2 * k];
        System.arraycopy(beforeCode, 0, ans, 0, beforeCode.length);

        if (showHint) {
            ans[beforeCode.length] = START_HINT;
            ans[beforeCode.length + 1] = START_DIRECTIVE;
            ans[beforeCode.length + text.length + 1] = END_DIRECTIVE;
            ans[beforeCode.length + text.length + 2] = END_HINT;
        } else {
            ans[beforeCode.length] = START_DIRECTIVE;
            ans[beforeCode.length + text.length] = END_DIRECTIVE;
        }
        System.arraycopy(text, 0, ans, beforeCode.length + k, text.length);
        System.arraycopy(afterCode, 0, ans, beforeCode.length + text.length + 2 * k, afterCode.length);

        return ans;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/utils/DirectivesUtils.kt
package org.stepik.core.utils

import com.intellij.openapi.application.ApplicationManager.getApplication
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.util.Computable
import com.intellij.openapi.vfs.VirtualFile
import org.stepik.core.SupportedLanguages
import java.util.*

private const val START_DIRECTIVE = "Stepik code: start"
private const val END_DIRECTIVE = "Stepik code: end"
private const val START_HINT = "Please note, only the code BELOW will be sent to Stepik.org"
private const val END_HINT = "Please note, only the code ABOVE will be sent to Stepik.org"

private const val MESSAGE = "Do you want to remove Stepik directives and external code?\n" + "You can undo this action using \"ctrl + Z\"."

fun getFileText(vf: VirtualFile): String {
    return getApplication().runReadAction(Computable {
        val document = FileDocumentManager.getInstance().getDocument(vf)
        return@Computable document?.text ?: ""
    })
}

fun getTextUnderDirectives(text: String, language: SupportedLanguages): String {
    val result = text.split("\n").toMutableList()
    val (start, end) = findDirectives(result, language)

    val commentPrefix = language.comment

    for (i in 0 until start) {
        result[i] = commentPrefix + result[i]
    }
    for (i in end + 1 until result.size) {
        result[i] = commentPrefix + result[i]
    }

    return result.joinToString("\n")
}

/**
 * Find first "Stepik code: start" and last "Stepik code: end". Return Pair(start, end).
 *
 * If "Stepik code: start" not found, start = -1
 *
 * If "Stepik code: end" not found, end = text.length
 */

fun findDirectives(text: String, language: SupportedLanguages): Pair<Int, Int> {
    return findDirectives(text.split("\n"), language)
}

fun findDirectives(text: List<String>, language: SupportedLanguages): Pair<Int, Int> {
    var start = -1
    var end = text.size
    val commentPrefixSize = language.comment.length

    for (i in text.indices) {
        val line = text[i]
        if (language.isCommentedLine(line) && isStart(line, commentPrefixSize)) {
            start = i
            break
        }
    }

    for (i in text.size - 1 downTo start + 1) {
        val line = text[i]
        if (language.isCommentedLine(line) && isEnd(line, commentPrefixSize)) {
            end = i
            break
        }
    }

    return start to end
}

private fun isStart(line: String, commentPrefixSize: Int): Boolean {
    return START_DIRECTIVE == line.trim().substring(commentPrefixSize).trim()
}

private fun isEnd(line: String, commentPrefixSize: Int): Boolean {
    return END_DIRECTIVE == line.trim().substring(commentPrefixSize).trim()
}

fun writeTextToFile(text: String, file: VirtualFile, project: Project) {
    val document = FileDocumentManager.getInstance().getDocument(file) ?: return

    CommandProcessor.getInstance()
            .executeCommand(project, {
                getApplication().runWriteAction { document.setText(text) }
            }, "Stepik directives process", "Stepik directives process")
}

fun removeAmbientCode(text: String, showHint: Boolean, language: SupportedLanguages, confirm: Boolean = true): String {
    val splittedText = text.split("\n")
    val size = splittedText.size
    val (start, end) = findDirectives(text, language)

    val k = if (showHint) 1 else 0
    val before = if (start > 0) splittedText.subList(0, start - k).joinToString("\n") else ""
    val e = end + k + 1
    val after = if (e < size) splittedText.subList(e, size).joinToString("\n") else ""

    if (confirm && (before != language.beforeCode || after != language.afterCode)) {
        val answer = Messages.showYesNoDialog(MESSAGE, "Information", Messages.getInformationIcon())
        if (answer != Messages.YES) {
            return splittedText.joinToString("\n")
        }
    }

    return splittedText.subList(start + 1, end).joinToString("\n")
}


fun StringBuilder.appendLnIf(string: String, condition: Boolean): StringBuilder {
    return if (condition) {
        appendln(string)
    } else {
        this
    }
}

fun insertAmbientCode(text: String, lang: SupportedLanguages, showHint: Boolean): String {
    return StringBuilder().appendln(lang.beforeCode ?: "")
            .appendLnIf(lang.comment(START_HINT), showHint)
            .appendln(lang.comment(START_DIRECTIVE))
            .appendln(text)
            .appendln(lang.comment(END_DIRECTIVE))
            .appendLnIf(lang.comment(END_HINT), showHint)
            .append(lang.afterCode ?: "")
            .toString()
            .trim()
}

fun replaceCode(text: String, code: String, language: SupportedLanguages): String {
    val textLines = text.split("\n")
    val codeLines = code.split("\n")
    val (start, end) = findDirectives(textLines, language)
    val beforeCodeSize = start + 1
    val afterCodeSize = textLines.size - end

    val resultLines = mutableListOf<String>()
    resultLines.addAll(textLines.subList(0, beforeCodeSize))
    resultLines.addAll(codeLines)
    resultLines.addAll(textLines.subList(end, end + afterCodeSize))

    return resultLines.joinToString("\n")
}

fun uncommentAmbientCode(text: String, language: SupportedLanguages): String {
    var lines = text.split("\n")
    val (start, end) = findDirectives(lines, language)
    val commentPrefix = language.comment
    val commentPrefixSize = commentPrefix.length
    for (i in 0 until start) {
        lines = uncomment(language, lines, commentPrefixSize, i)
    }
    for (i in end + 1 until lines.size) {
        lines = uncomment(language, lines, commentPrefixSize, i)
    }

    return lines.joinToString("\n")
}

private fun uncomment(language: SupportedLanguages,
                      lines: List<String>,
                      commentPrefixSize: Int,
                      lineIndex: Int): List<String> {
    val result = ArrayList(lines)
    if (language.isCommentedLine(lines[lineIndex])) {
        result[lineIndex] = lines[lineIndex].substring(commentPrefixSize)
    }

    return result
}

fun containsDirectives(text: String, language: SupportedLanguages): Boolean {
    val splittedText = text.split("\n")
    val commentPrefixSize = language.comment.length

    for (i in splittedText.indices) {
        var line = splittedText[i]
        if (language.isCommentedLine(line)) {
            line = splittedText[i].trim().substring(commentPrefixSize).trim()
            if (START_DIRECTIVE == line || END_DIRECTIVE == line) {
                return true
            }
        }
    }

    return false
}
